src/protocol/get_cert.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! `GetCert` request and response.
+//!
+//! This module provides a Cerberus command for requesting certificates.
+
+use crate::io::Read;
+use crate::io::Write;
+use crate::mem::Arena;
+use crate::mem::ArenaExt as _;
+use crate::protocol::wire::FromWire;
+use crate::protocol::wire::FromWireError;
+use crate::protocol::wire::ToWire;
+use crate::protocol::wire::ToWireError;
+use crate::protocol::Command;
+use crate::protocol::CommandType;
+use crate::protocol::Request;
+use crate::protocol::Response;
+
+#[cfg(feature = "arbitrary-derive")]
+use libfuzzer_sys::arbitrary::{self, Arbitrary};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+/// A command for requesting a chunk of a certificate.
+///
+/// Corresponds to [`CommandType::GetCert`].
+pub enum GetCert {}
+
+impl<'a> Command<'a> for GetCert {
+    type Req = GetCertRequest;
+    type Resp = GetCertResponse<'a>;
+}
+
+/// The [`GetCert`] request.
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+#[cfg_attr(feature = "arbitrary-derive", derive(Arbitrary))]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct GetCertRequest {
+    /// The slot number of the chain to read from.
+    pub slot: u8,
+    /// The number of the cert to request, indexed from the root.
+    pub cert_number: u8,
+    /// The offset in bytes from the start of the certificate to read from.
+    pub offset: u16,
+    /// The number of bytes to read.
+    pub len: u16,
+}
+make_fuzz_safe!(GetCertRequest);
+
+impl Request<'_> for GetCertRequest {
+    const TYPE: CommandType = CommandType::GetCert;
+}
+
+impl<'a> FromWire<'a> for GetCertRequest {
+    fn from_wire<R: Read, A: Arena>(
+        mut r: R,
+        _: &'a A,
+    ) -> Result<Self, FromWireError> {
+        let slot = r.read_le()?;
+        let cert_number = r.read_le()?;
+        let offset = r.read_le()?;
+        let len = r.read_le()?;
+        Ok(Self {
+            slot,
+            cert_number,
+            offset,
+            len,
+        })
+    }
```

Comment:
Is there a need for these `let` bindings?  Could you not incorporate the `r.read_le()`s into the initializer directly?

Reply:
To my knowledge, Rust doesn't specify the order in which the arguments to a struct literal are evaluated. If it turns out I'm wrong, I can do a mass refactor of all the parsers, since they all do this.